File: apps/native/ios/App/App/RavenMediaPlugin.swift
import Foundation
import Capacitor

/// Session for the media handler, and the media/ cache trim (contract: apps/web/src/native/media.ts).
@objc(RavenMediaPlugin)
public class RavenMediaPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "RavenMediaPlugin"
    public let jsName = "RavenMedia"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trim", returnType: CAPPluginReturnPromise),
    ]
    private var folder: URL { FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0].appendingPathComponent("media") }

    @objc func setSession(_ call: CAPPluginCall) {
        let origin = call.getString("origin") ?? "", token = call.getString("token") ?? ""
        // The handler's state is main-thread only.
        DispatchQueue.main.async {
            RavenMediaHandler.shared.origin = origin
            RavenMediaHandler.shared.token = token
            call.resolve()
        }
    }

    // Nothing goes while the folder fits the budget. Over it, folders outside keep go first (oldest
    // first), then keep from its end: keep is newest message first.
    @objc func trim(_ call: CAPPluginCall) {
        // Folder names are base-36 hashes; anything else would be a path outside the cache.
        let keep = (call.getArray("keep", String.self) ?? []).filter { $0.range(of: "^[a-z0-9]+$", options: .regularExpression) != nil }
        let budget = Int64(call.getDouble("budgetBytes") ?? Double(Int64.max))
        let fm = FileManager.default
        guard let folders = try? fm.contentsOfDirectory(at: folder, includingPropertiesForKeys: [.contentModificationDateKey]) else { return call.resolve() }
        let kept = Set(keep)
        var total = folders.reduce(Int64(0)) { $0 + size($1) }
        let modified = { (url: URL) in (try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast }
        let others = folders.filter { !kept.contains($0.lastPathComponent) }.sorted { modified($0) < modified($1) }
        for url in others where total > budget {
            total -= size(url)
            try? fm.removeItem(at: url)
        }
        for name in keep.reversed() where total > budget {
            let url = folder.appendingPathComponent(name)
            total -= size(url)
            try? fm.removeItem(at: url)
        }
        call.resolve()
    }

    private func size(_ url: URL) -> Int64 {
        let files = (try? FileManager.default.contentsOfDirectory(at: url, includingPropertiesForKeys: [.fileSizeKey])) ?? []
        return files.reduce(0) { $0 + Int64((try? $1.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0) }
    }
}

File: apps/native/android/app/src/main/java/raven/thecommit/company/RavenMediaPlugin.java
package raven.thecommit.company;

import com.getcapacitor.JSArray;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/** Session for the media handler, and the media/ cache trim (contract: apps/web/src/native/media.ts). */
@CapacitorPlugin(name = "RavenMedia")
public class RavenMediaPlugin extends Plugin {
    @PluginMethod
    public void setSession(PluginCall call) {
        RavenMediaHandler.origin = call.getString("origin", "");
        RavenMediaHandler.token = call.getString("token", "");
        call.resolve();
    }

    // Nothing goes while the folder fits the budget. Over it, folders outside keep go first (oldest
    // first), then keep from its end: keep is newest message first.
    @PluginMethod
    public void trim(PluginCall call) {
        JSArray keepArray = call.getArray("keep", new JSArray());
        // getLong only accepts a JSON long; a budget that fits an int would read as missing.
        Object rawBudget = call.getData().opt("budgetBytes");
        long budget = rawBudget instanceof Number ? ((Number) rawBudget).longValue() : Long.MAX_VALUE;
        List<String> keep = new ArrayList<>();
        // Folder names are base-36 hashes; anything else would be a path outside the cache.
        try {
            for (int i = 0; i < keepArray.length(); i++) if (keepArray.getString(i).matches("[a-z0-9]+")) keep.add(keepArray.getString(i));
        } catch (Exception ignored) { }
        File media = new File(getContext().getCacheDir(), "media");
        File[] folders = media.listFiles();
        if (folders == null) { call.resolve(); return; }
        Set<String> kept = new HashSet<>(keep);
        long total = 0;
        List<File> others = new ArrayList<>();
        for (File folder : folders) {
            total += size(folder);
            // Notification avatars go last of all, however old: they are small and re-read constantly.
            if (!kept.contains(folder.getName()) && !AvatarCache.FOLDER.equals(folder.getName())) others.add(folder);
        }
        others.sort((a, b) -> Long.compare(a.lastModified(), b.lastModified()));
        for (File folder : others) {
            if (total <= budget) break;
            total -= size(folder);
            delete(folder);
        }
        for (int i = keep.size() - 1; i >= 0 && total > budget; i--) {
            File folder = new File(media, keep.get(i));
            total -= size(folder);
            delete(folder);
        }
        // The faces are reached only after every other cached file has been deleted.
        File avatars = new File(media, AvatarCache.FOLDER);
        if (total > budget) delete(avatars);
        call.resolve();
    }

    private static long size(File file) {
        if (!file.isDirectory()) return file.length();
        long total = 0;
        File[] children = file.listFiles();
        if (children != null) for (File child : children) total += size(child);
        return total;
    }

    private static void delete(File file) {
        File[] children = file.listFiles();
        if (children != null) for (File child : children) delete(child);
        //noinspection ResultOfMethodCallIgnored
        file.delete();
    }
}
